nerium/main.py

```diff
@@ -13,69 +14,135 @@
+    subclass to get results from.
 
-    Usage:
-        Subclass and provide a 'result' method
+    Provides a get_query method to retrieve path and result_cls attributes
+    from a query_name
     """
 
-    def __init__(self, query_name, **kwargs):
-        self.query_name = query_name
-        self.kwargs = kwargs
+    # TODO: Provide for registration from frontmatter or direct config
 
-    def get_query_path(self):
+    def __init__(self):
+        pass
+
+    def lookup_extension(self, _key):
+        # TODO: Define in external config file
+        query_extension_lookup = {
+            'sql': 'SQLResultSet',
+        }
+        try:
+            return query_extension_lookup[_key]
+        except KeyError:
+            return 'SQLResultSet'
+
+    def register_paths(self):
         query_directory = Path(os.getenv('QUERY_PATH', 'query_files'))
         flat_directory = list(query_directory.glob('**/*'))
-        files = [i for i in flat_directory if i.is_file()]
+        registry = [
+            dict(
+                query_name=i.stem,
+                query_path=i,
+                result_cls=self.lookup_extension(i.suffix))
+            for i in flat_directory if i.is_file()
+        ]
+        return registry
+
+    def get_query(self, query_name):
         try:
-            query_path = next(
-                i for i in sorted(files) if i.stem == self.query_name)
-            return query_path
+            query = next(
+                i for i in self.register_paths()
+                if query_name == i['query_name'])
+            return query
         except StopIteration:
             return None
 
+
+class Query():
+    """ Finds query in registry, and looks up ResultSet subclass from
+    path file extension. Submits query_path to ResultSet and returns results
+    """
+
+    def __init__(self, query_name, **kwargs):
+        self.query_name = query_name
+        self.kwargs = kwargs
+
     def result_set(self):
-        """ if your functionResultSet subclass needs no script,
-            simply return self.result() here.
-        """
-        if not self.get_query_path():
-            return [
-                {
-                    'error': "No query found matching {}".format(
-                        self.query_name)
-                }
-            ]
+        registry = QueryRegistry()
+        query = registry.get_query(self.query_name)
+        if not query:
+            return [{
+                'error':
+                "No query found matching {}".format(self.query_name)
+            }]
         else:
-            return self.result()
+            query_path = query['query_path']
+            result_mods = import_module('nerium.contrib.resultset')
+            result_cls = getattr(result_mods, query['result_cls'])
+            loader = result_cls(query_path, **self.kwargs)
+            query_result = loader.result()
+            return query_result
```

Comment:
maybe return query_result and kwargs here for processing in the formatter, but we might want to also include other metadata from the frontmatter. :)